transport/src/main/java/io/netty/bootstrap/DynamicLocalSocketAddress.java

```diff
@@ -0,0 +1,123 @@
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.NetworkInterface;
+import java.net.SocketAddress;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Iterator;
+
+/**
+ * {@link SocketAddress} which will provide the local {@link SocketAddress} to bind to after the remote address
+ * is resolved and so can be used with {@link Bootstrap#connect(SocketAddress, SocketAddress)}.
+ */
+public abstract class DynamicLocalSocketAddress extends SocketAddress {
+
+    /**
+     * Return a {@link SocketAddress} to bind to for the given {@code remote} {@link SocketAddress} or {@code null}
+     * if no specific should be used.
+     */
+    public abstract SocketAddress address(SocketAddress remote);
+
+    public static DynamicLocalSocketAddress networkInterface(NetworkInterface networkInterface, int port) {
+        return new DynamicInetSocketAddress(networkInterface, port);
+    }
+
+    public static DynamicLocalSocketAddress any(Iterable<? extends InetSocketAddress> iterable) {
+        return new DynamicInetSocketAddress(iterable);
+    }
+
+    public static DynamicLocalSocketAddress either(InetSocketAddress inet4Address, InetSocketAddress inet6Address) {
+        if (!(inet4Address.getAddress() instanceof Inet4Address)) {
+            throw new IllegalArgumentException();
+        }
+        if (!(inet6Address.getAddress() instanceof Inet6Address)) {
+            throw new IllegalArgumentException();
+        }
+
+        return new DynamicInetSocketAddress(Arrays.asList(inet4Address, inet6Address));
+    }
+
+    private static final class DynamicInetSocketAddress extends DynamicLocalSocketAddress {
+        private final Iterable<? extends InetSocketAddress> iterable;
+
+        DynamicInetSocketAddress(final NetworkInterface networkInterface, final int port) {
+            this(new Iterable<InetSocketAddress>() {
+                @Override
+                public Iterator<InetSocketAddress> iterator() {
+                    return new EnumerationIterator(networkInterface.getInetAddresses(), port);
+                }
+            });
+            ObjectUtil.checkNotNull(networkInterface, "networkInterface");
+        }
+
+        DynamicInetSocketAddress(Iterable<? extends InetSocketAddress> iterable) {
+            this.iterable = ObjectUtil.checkNotNull(iterable, "iterable");
+        }
+
+        @Override
+        public SocketAddress address(SocketAddress remote) {
+            if (!(remote instanceof InetSocketAddress)) {
+                throw new IllegalArgumentException("Only InetSocketAddress is supported");
+            }
+
+            return inetSocketAddress0(((InetSocketAddress) remote).getAddress().getClass());
+        }
+
+        private InetSocketAddress inetSocketAddress0(Class<? extends InetAddress> clazz) {
+            for (InetSocketAddress address: iterable) {
+                if (clazz.isInstance(address.getAddress())) {
```

Comment:
At least in our case, just choosing the first address in `NetworkInterface.getInetAddresses()` that matches the class of the remote address would lead to some issues. Some interfaces that we might configure in our Netty application have both a link-global and global scope IPv6 address, for example. That's why we filter the IP addresses in the interface to those that are not link-local or loopback addresses in the snippet I provided in #8940.

Maybe, in special cases like this, we should still extract the relevant IPv4 and IPv6 addresses manually in our own code and pass them to the `either(inet4Address, inet6Address)` method. Or maybe the `networkInterface(networkInterface, port)` should take one or more arguments that configure which types of IP addresses should be used?

Reply:
That’s why I made it an abstract class which means you could just extend it . Does this not work ?

Reply:
Ah yes, that would work. I just wonder if the case of having multiple addresses (e.g. both link-local and global scope for IPv6) is prevalent enough to warrant accounting for it in the default implementation.

Reply:
honestly I have no idea what would be the best default implementation here. Maybe we should just not include any implementation at all but just provide the abstract class

Reply:
I just removed all the impls...